### ISSUE TITLE
FlexAttention FLASH ignores sparse BlockMask metadata when mask_mod is trivial

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#188322]
#### Human Note

I had my agent generate this small issue for a real bug I found and will fix :)

### Agent report
`torch.nn.attention.flex_attention` with `kernel_options={"BACKEND": "FLASH"}` appears to mishandle a sparse `BlockMask` when sparsity is represented only by `kv_*` / `full_kv_*` metadata and the `mask_mod` is trivial/noop.

The same `BlockMask`:

- has the expected `to_dense()` representation,
- produces correct output with `BACKEND="TRITON"`, and
- produces correct output with `BACKEND="FLASH"` if I attach an equivalent nontrivial tile-level `mask_mod`.

This suggests the Flex FLASH lowering/CuTe backend may be deciding whether to honor sparse block metadata from whether the `mask_mod` graph is trivial, rather than from the presence of sparse `BlockMask` metadata.

### Repro

```python
from functools import partial
import math

import torch
from torch.nn.attention.flex_attention import BlockMask, flex_attention


def run(name, fn, q, k, v, block_mask, ref):
    try:
        out = fn(q, k, v, block_mask=block_mask)
        torch.cuda.synchronize()
        diff = (out.float() - ref.float()).abs()
        ok = torch.allclose(out, ref, atol=5e-2, rtol=5e-2)
        print(f"{name}: ok={ok} max={diff.max().item():.6f} mean={diff.mean().item():.6f}")
    except Exception as exc:
        print(f"{name}: ERROR {type(exc).__name__}: {str(exc).splitlines()[0]}")


def main():
    print("torch", torch.__version__)
    print("device", torch.cuda.get_device_name(), torch.cuda.get_device_capability())

    torch.manual_seed(0)
    batch, heads, q_blocks, kv_blocks, block_size, head_dim, top_k = 1, 2, 4, 4, 256, 128, 2
    q = torch.randn(batch, heads, q_blocks * block_size, head_dim, device="cuda", dtype=torch.bfloat16)
    k = torch.randn(batch, heads, kv_blocks * block_size, head_dim, device="cuda", dtype=torch.bfloat16)
    v = torch.randn(batch, heads, kv_blocks * block_size, head_dim, device="cuda", dtype=torch.bfloat16)

    selected = torch.tensor(
        [[[[0, 2], [1, 3], [0, 3], [2, 3]], [[0, 2], [1, 3], [0, 3], [2, 3]]]],
        device="cuda",
        dtype=torch.int32,
    )
    num_blocks = torch.full(selected.shape[:-1], top_k, device="cuda", dtype=torch.int32)
    zero_num_blocks = torch.zeros_like(num_blocks)
    zero_indices = torch.zeros(*selected.shape[:-1], kv_blocks, device="cuda", dtype=torch.int32)
    full_indices = torch.zeros_like(zero_indices)
    full_indices[..., :top_k] = selected

    block_map = torch.zeros(batch, heads, q_blocks, kv_blocks, device="cuda", dtype=torch.bool)
    block_map.scatter_(-1, selected.long(), True)
    dense_mask = block_map.repeat_interleave(block_size, -2).repeat_interleave(block_size, -1)
    ref = torch.softmax((q @ k.transpose(-2, -1) / math.sqrt(head_dim)).masked_fill(~dense_mask, -float("inf")), dim=-1) @ v

    full_metadata_only = BlockMask.from_kv_blocks(
        kv_num_blocks=zero_num_blocks,
        kv_indices=zero_indices,
        full_kv_num_blocks=num_blocks,
        full_kv_indices=full_indices,
        BLOCK_SIZE=block_size,
        seq_lengths=(q_blocks * block_size, kv_blocks * block_size),
    )

    partial_metadata_only = BlockMask.from_kv_blocks(
        kv_num_blocks=num_blocks,
        kv_indices=full_indices,
        BLOCK_SIZE=block_size,
        seq_lengths=(q_blocks * block_size, kv_blocks * block_size),
    )

    def mask_mod(b, h, q_idx, kv_idx):
        return block_map[b, h, q_idx // block_size, kv_idx // block_size]

    full_metadata_with_mask_mod = BlockMask.from_kv_blocks(
        kv_num_blocks=zero_num_blocks,
        kv_indices=zero_indices,
        full_kv_num_blocks=num_blocks,
        full_kv_indices=full_indices,
        BLOCK_SIZE=block_size,
        seq_lengths=(q_blocks * block_size, kv_blocks * block_size),
        mask_mod=mask_mod,
    )

    triton = torch.compile(partial(flex_attention, kernel_options={"BACKEND": "TRITON"}), dynamic=False)
    flash = torch.compile(partial(flex_attention, kernel_options={"BACKEND": "FLASH"}), dynamic=False)

    print("dense block mask")
    print(full_metadata_only.to_dense().cpu())
    run("TRITON full_kv metadata only", triton, q, k, v, full_metadata_only, ref)
    run("FLASH full_kv metadata only", flash, q, k, v, full_metadata_only, ref)
    run("TRITON partial kv metadata only", triton, q, k, v, partial_metadata_only, ref)
    run("FLASH partial kv metadata only", flash, q, k, v, partial_metadata_only, ref)
    run("TRITON full_kv metadata + mask_mod", triton, q, k, v, full_metadata_with_mask_mod, ref)
    run("FLASH full_kv metadata + mask_mod", flash, q, k, v, full_metadata_with_mask_mod, ref)


if __name__ == "__main__":
    main()
```

### Output

```text
torch 2.14.0.dev20260625+cu132
device NVIDIA GB200 (10, 0)
dense block mask
tensor([[[[1, 0, 1, 0],
          [0, 1, 0, 1],
          [1, 0, 0, 1],
          [0, 0, 1, 1]],

         [[1, 0, 1, 0],
          [0, 1, 0, 1],
          [1, 0, 0, 1],
          [0, 0, 1, 1]]]], dtype=torch.int32)
TRITON full_kv metadata only: ok=True max=0.005859 mean=0.000308
FLASH full_kv metadata only: ok=False max=0.258789 mean=0.040552
TRITON partial kv metadata only: ok=True max=0.005859 mean=0.000308
FLASH partial kv metadata only: ok=False max=0.258789 mean=0.040552
TRITON full_kv metadata + mask_mod: ok=True max=0.005859 mean=0.000308
FLASH full_kv metadata + mask_mod: ok=True max=0.005859 mean=0.000309
```

### Expected behavior

`BACKEND="FLASH"` should honor sparse `BlockMask` metadata (`kv_*` and/or `full_kv_*`) even when the `mask_mod` graph is trivial/noop. It should match the dense masked reference and the `BACKEND="TRITON"` result.

### Additional context

In `torch/_inductor/kernel/flex/flex_flash_attention.py`, the FLASH lowering computes:

```python
mask_graph_is_trivial = is_trivial_mask_graph(mask_graph.graph_module)
needs_block_mask = not mask_graph_is_trivial
has_full_blocks = full_kv_num_blocks is not None
```

and then passes `NEEDS_BLOCK_MASK=needs_block_mask` to the CuTeDSL template. In the failing case, sparse full-block metadata is present (`has_full_blocks=True`) but `needs_block_mask=False` because the mask graph is trivial. Attaching an equivalent nontrivial `mask_mod` makes `needs_block_mask=True`, and the result becomes correct.

So the failure may be that Flex FLASH uses the mask graph to decide whether block-sparse metadata is needed, rather than treating sparse `BlockMask` metadata as sufficient to require block-mask handling.

### Versions

- PyTorch: `2.14.0.dev20260625+cu132`
- CUDA runtime/toolkit on the test machine: CUDA 13.x
- GPU: NVIDIA GB200, capability `(10, 0)`
- dtype: `torch.bfloat16`

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @Chillee @yanboliang @BoyuanFeng @liangel-02 @howardzhang-cv
